tests/align/align_create_tensor_torch.py

```diff
@@ -0,0 +1,646 @@
+
+    if (operator_name in param_weight_op):
+        torch.save(weight.grad.cpu(), os.path.join(
+            OUT_DIR, "torch_weight_grad.pt"))
+    if (operator_name in param_bias_op):
+        torch.save(bias.grad.cpu(), os.path.join(
+            OUT_DIR, "torch_bias_grad.pt"))
+
+# run tests for all operators
+
+
+def create_tensors_for_add_torch():
+    inp1: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    inp2: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    label: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    output = torch.add(
+        input=inp1,
+        other=inp2
+    ).to(DEVICE)
+    output.requires_grad = True
+    return label, output
+
+
+def create_tensors_for_concat_torch():
+    inp1: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    inp2: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    inp3: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    label: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, SEQ_LENGTH * 3, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    output = torch.cat(
+        tensors=(inp1, inp2, inp3),
+        dim=1
+    ).to(DEVICE)
+    output.requires_grad = True
+
+    return label, output
+
+
+def create_tensors_for_conv2d_torch(param_dir):
+    KERNEL_SIZE = 3
+    IN_CHANNELS = 3
+    OUTPUT_SIZE = 510
+    OUT_CHANNELS = 5
+    conv2d = torch.nn.Conv2d(
+        in_channels=IN_CHANNELS,
+        out_channels=OUT_CHANNELS,
+        kernel_size=KERNEL_SIZE
+    ).to(DEVICE)
+
+    linear_weight = torch.load(os.path.join(param_dir, "ff_weight.pt"))
+    linear_bias = torch.load(os.path.join(param_dir, "ff_bias.pt"))
+    assert conv2d.weight.shape == linear_weight.shape, (
+        "Shape mismatch: " f"FF={linear_weight.shape} torch={conv2d.weight.shape}"
+    )
+    assert conv2d.bias.shape == linear_bias.shape, (
+        "Shape mismatch: " f"FF={linear_bias.shape} torch={conv2d.bias.shape}"
+    )
+
+    conv2d.weight = torch.nn.Parameter(linear_weight.to(DEVICE))
+    conv2d.bias = torch.nn.Parameter(linear_bias.to(DEVICE))
+
+    # generate input/label tensors
+    # imitating 3-channel image input
+    inp: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, 3, INPUT_SIZE, INPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+    label: torch.Tensor = gen_tensor(
+        (BATCH_SIZE, 5, OUTPUT_SIZE, OUTPUT_SIZE),
+        dtype="float32"
+    ).to(DEVICE)
+
+    output = conv2d(inp)
+    conv2d.zero_grad()
+
+    return label, output, conv2d.weight, conv2d.bias
+
+
+def create_tensors_for_cos_torch():
```

Comment:
All this tensor code seems duplicated between ff and torch. Is there a reason for this or could the redundancy be eliminated? Seems like the tensor should always be the same shape, but correct me if I'm wrong here

Reply:
yes, some of them are identical, some are not. thanks for reminding. solved by created function in utils